File: WindowsCE/NurSample_vs2008/NurSample/HHUtils.cs
/*=======================================================================================

    Hand Held Utils for Nordic ID devices like Morphic and Merlin

=======================================================================================*/
using System;
using System.Collections.Generic;
using System.Text;
using Microsoft.Win32;
using NordicId;

namespace NurSample
{
    static class HHUtils
    {
        static private int lastKeepAliveTick = 0;
        static private IntPtr userActivityEvent = IntPtr.Zero;

        /// <summary>
        /// Keeps the device alive.
        /// </summary>
        static public void KeepDeviceAlive()
        {
            // Prevent unnecessary CPU load
            int tick = Environment.TickCount;
            if (tick - lastKeepAliveTick < 5000)
                return;
            lastKeepAliveTick = tick;

            // Create ActivityEvent if not exist
            if (userActivityEvent == IntPtr.Zero)
            {
                using (RegistryKey key = Registry.LocalMachine.OpenSubKey("System\\GWE"))
                {
                    object value = key.GetValue("ActivityEvent");
                    key.Close();
                    if (value == null) return;
                    string activityEventName = (string)value;
                    userActivityEvent = WIN32.CreateEvent(IntPtr.Zero, false, false, activityEventName);
                }
            }

            // Signal user activity to the GWE and Backlight driver
            if (userActivityEvent != IntPtr.Zero)
                WIN32.SetEvent(userActivityEvent);
        }
    }
}

File: WindowsCE/NurSample_vs2008_VB/NurSampleCE_VB/HHUtils.vb
'=======================================================================================
'
'    Hand Held Utils for Nordic ID devices like Morphic and Merlin
'
'=======================================================================================

Imports System.Collections.Generic
Imports System.Text
Imports Microsoft.Win32
Imports NordicId

NotInheritable Class HHUtils
    Private Sub New()
    End Sub
    Private Shared lastKeepAliveTick As Integer = 0
    Private Shared userActivityEvent As IntPtr = IntPtr.Zero

    ''' <summary>
    ''' Keeps the device alive.
    ''' </summary>
    Public Shared Sub KeepDeviceAlive()
        ' Prevent unnecessary CPU load
        Dim tick As Integer = Environment.TickCount
        If tick - lastKeepAliveTick < 5000 Then
            Return
        End If
        lastKeepAliveTick = tick

        ' Create ActivityEvent if not exist
        If userActivityEvent = IntPtr.Zero Then
            Using key As RegistryKey = Registry.LocalMachine.OpenSubKey("System\GWE")
                Dim value As Object = key.GetValue("ActivityEvent")
                key.Close()
                If value Is Nothing Then
                    Return
                End If
                Dim activityEventName As String = DirectCast(value, String)
                userActivityEvent = WIN32.CreateEvent(IntPtr.Zero, False, False, activityEventName)
            End Using
        End If

        ' Signal user activity to the GWE and Backlight driver
        If userActivityEvent <> IntPtr.Zero Then
            WIN32.SetEvent(userActivityEvent)
        End If
    End Sub
End Class
